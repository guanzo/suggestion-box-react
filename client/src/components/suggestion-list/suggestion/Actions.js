import React, { Component } from 'react';
import Upvote from './Upvote'

class Actions extends Component {
    render() {
        
        return (
            <div class="suggestion-actions">
                <div class="flex">
                    <Upvote {...this.props}></Upvote>
                </div>
            </div>
        )
    }
}

export default Actions;
