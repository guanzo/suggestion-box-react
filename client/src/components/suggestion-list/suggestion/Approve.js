import React, { Component } from 'react';
import { connect } from 'react-redux'
import { approveSuggestion } from '@/store/suggestions-admin'
const { STATUS_APPROVED } = require('@shared/suggestion-util')

class Approve extends Component {
    render() {
		let { status } = this.props
		let component;
		if(status === STATUS_APPROVED)
			component = <div class="has-text-success help">approved</div>
		else
			component = <i class='fa fa-check-square-o' 
							onClick={this.props.approveSuggestion}
						></i>

        return (
            <div class="m-l-a">
                { component }
            </div>
        )
	}
}

const mapDispatchToProps = (dispatch,ownProps) => {
    return {
        approveSuggestion: ()=> dispatch(approveSuggestion(ownProps))
    }
}
const Approve_C = connect(null, mapDispatchToProps)(Approve)

export default Approve_C;

