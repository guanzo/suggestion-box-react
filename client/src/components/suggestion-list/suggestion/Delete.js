import React, { Component } from 'react';
import { connect } from 'react-redux'
import { deleteSuggestion } from '@/store/suggestions'
const { STATUS_DELETED } = require('@shared/suggestion-util')

class Delete extends Component {
	constructor(){
		super()
		this.state = {
			askToConfirm: false
		}
	}
    render() {
		let { askToConfirm } = this.state
		let { status } = this.props
		let component;
		if(status === STATUS_DELETED)
			component = <div class="has-text-danger help">deleted</div>
		else if(askToConfirm)
			component = this.confirm()
		else
			component = <i class='fa fa-trash-o' onClick={e=>this.setState({ askToConfirm: true })}></i>

        return (
            <div class="m-l-a">
                { component }
            </div>
        )
	}
	confirm(){
		const style = { cursor: 'pointer' }
		return (
			<div class="help flex wrap justify-end m-t-0">
				<span class="has-text-danger">are you sure?</span>&nbsp;
				<div class="flex">
					<div onClick={this.props.deleteSuggestion} style={style}>yes</div>&nbsp;
					<div class="has-text-danger">/</div>&nbsp;
					<div onClick={e=>this.setState({ askToConfirm: false })} style={style}>no</div>
				</div>	
			</div> 
		)
	}
}

const mapDispatchToProps = (dispatch,ownProps) => {
    return {
        deleteSuggestion: ()=> dispatch(deleteSuggestion(ownProps))
    }
}
const Delete_C = connect(null, mapDispatchToProps)(Delete)

export default Delete_C;

