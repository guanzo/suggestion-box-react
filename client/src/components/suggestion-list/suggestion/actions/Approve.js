import React, { PureComponent } from 'react';
import { connect } from 'react-redux'
import { updateSuggestionStatus } from '@/store/suggestions-admin'
const { STATUS_APPROVED, LIST_APPROVED } = require('@shared/suggestion-util')

//admins can only approve suggestions in the pending list
class Approve extends PureComponent {
    render() {
		let { status, listType } = this.props
		if(listType === LIST_APPROVED)
			return null

		let component;
		if(status === STATUS_APPROVED)
			component = <div className="has-text-success is-size-7">approved</div>
		else
			component = <span className="approve-suggestion icon  has-text-success">
							<i className='fa fa-check-square' 
								onClick={this.props.approveSuggestion}
							></i>
						</span>

        return component
	}
}

const mapDispatchToProps = (dispatch,ownProps) => {
    return {
        approveSuggestion: ()=> {
			let data = {
				...ownProps,
				status: STATUS_APPROVED
			}
			dispatch(updateSuggestionStatus(data))
		}
    }
}
export default connect(null, mapDispatchToProps)(Approve)


