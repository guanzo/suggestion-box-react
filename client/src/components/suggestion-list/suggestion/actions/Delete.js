/*eslint react/jsx-no-bind:0*/

import React, { Component } from 'react';
import { connect } from 'react-redux'
import { updateSuggestionStatus } from '@store/suggestions-admin'
const { STATUS_DELETED } = require('@shared/suggestion-util')

//admins can delete a suggestion in any list
class Delete extends Component {
	constructor(){
		super()
		this.state = {
			askToConfirm: false
		}
	}

    render() {
		let { askToConfirm } = this.state
		let { status } = this.props
		let component;
		if(status === STATUS_DELETED)
			component = <div className="has-text-danger is-size-7">deleted</div>
		else if(askToConfirm)
			component = this.confirm()
		else
			component = <span className="delete-suggestion icon has-text-danger">
							<i className='fa fa-trash-o' 
								onClick={e=>this.setState({ askToConfirm: true })}
							></i>
						</span>

        return component
	}
	confirm(){
		const style = { cursor: 'pointer' }
		return (
			<div className="flex wrap justify-end is-size-7">
				<span className="has-text-danger">are you sure?</span>&nbsp;
				<div className="flex">
					<div onClick={this.props.deleteSuggestion} style={style}>yes</div>&nbsp;
					<div className="has-text-danger">/</div>&nbsp;
					<div onClick={e=>this.setState({ askToConfirm: false })} style={style}>no</div>
				</div>	
			</div> 
		)
	}
}

const mapDispatchToProps = (dispatch,ownProps) => {
    return {
        deleteSuggestion: ()=> {
			let data = {
				...ownProps,
				status: STATUS_DELETED
			}
			dispatch(updateSuggestionStatus(data))
		}
    }
}
export default connect(null, mapDispatchToProps)(Delete)


