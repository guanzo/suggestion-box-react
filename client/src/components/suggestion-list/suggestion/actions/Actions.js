import React, { PureComponent } from 'react';
import Upvote from './Upvote'
import Emote from './emote/Emote'
import Approve from './Approve'
import Delete from './Delete'

class Actions extends PureComponent {

    render() {
		let { currentUser } = this.props
        return (
            <div className="suggestion-actions">
				<div className="flex">
					<Upvote {...this.props}></Upvote>
					<Emote></Emote>
					{ currentUser.isAdmin ? this.adminActions() : null }
				</div>
            </div>
        )
	}
	adminActions(){
		return (
			<div className="suggestion-admin-actions flex-center m-l-a">
				<Approve {...this.props}></Approve>
				<Delete {...this.props}></Delete> 
			</div>
		)
	}
}

export default Actions;
