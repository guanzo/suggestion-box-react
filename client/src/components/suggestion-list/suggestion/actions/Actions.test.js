import React from 'react';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Actions from './Actions';
import Approve from './Approve'
import Delete from './Delete'

const data = {
	hasUpvoted: true,
	votesLength: 1,
	broadcasterUpvoted: true,
	channel:{
		channelName: 'guanzo'
	},
	currentUser: {
		isAdmin: false
	}
}
function mockStore(store){
	let rootReducer = ()=>store
	return createStore(rootReducer)
}

describe('<Actions />',()=>{
	it('does not show admin actions to non-admins', () => {
		const store =mockStore(Object.assign({},data,{ currentUser:{ isAdmin:false } }))
		const wrapper = shallow(<Actions {...store.getState()} store={store}/>)
		expect(wrapper.find(Approve)).toHaveLength(0);
	})

	it('shows admin actions to admins', () => {
		const store = mockStore(Object.assign({},data,{ currentUser:{ isAdmin:true } }))
		const wrapper = shallow(<Actions {...store.getState()} store={store}/>)
		expect(wrapper.find(Approve)).toHaveLength(1);
	})
	it('renders correctly',()=>{
		const store = mockStore(Object.assign({},data,{ currentUser:{ isAdmin:true } }))
		const tree = renderer
		.create(<Actions {...store.getState()} store={store}/>)
		.toJSON();
	  	expect(tree).toMatchSnapshot();
	})
})