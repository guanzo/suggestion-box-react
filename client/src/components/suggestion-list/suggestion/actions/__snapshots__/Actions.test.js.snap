// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Actions /> renders correctly 1`] = `
<div
  className="suggestion-actions flex"
>
  <div
    className="flex align-center justify-start"
  >
    <span
      className="icon"
    >
      <i
        className="fa fa-thumbs-up has-text-primary"
        onClick={undefined}
        title=""
      />
    </span>
    <div
      className="is-size-7"
    >
      1
    </div>
    <p
      className="m-l-10 is-size-7"
    >
      guanzo likes this
    </p>
  </div>
  <div
    className="suggestion-admin-actions flex-center m-l-a"
  >
    <span
      className="approve-suggestion icon  has-text-success"
    >
      <i
        className="fa fa-check-square"
        onClick={[Function]}
      />
    </span>
    <span
      className="delete-suggestion icon has-text-danger"
    >
      <i
        className="fa fa-trash-o"
        onClick={[Function]}
      />
    </span>
  </div>
</div>
`;
